### PR TITLE
fix: close and join the pool on exit

### PR DESCRIPTION
**Pull request by Midnighter**
* [ ] fix #(issue number)
* [ ] description of feature/fix
* [ ] tests added/passed
* [ ] add an entry to the [next release](../release-notes/next-release.md)

### CODE REVIEW

**Comment by synchon** [comment]
Looks like the problem with CI is still there. Maybe re-run the pipeline for the ones that failed and see if that helps?

**Comment by Midnighter** [comment]
Funny that it's the Ubuntu ones that timed out this time rather than the Mac jobs.

**Comment by synchon** [comment]
Yes that's the weird part and makes me believe that either our time limits are too strict as @cdiener points out in #1187 , or something funny is going on Github's workflow management end.

File: src/cobra/util/process_pool.py
Comment (cdiener) [inline src/cobra/util/process_pool.py:85]: Don't you need to get the result before you close and join?

Comment (cdiener) [inline src/cobra/util/process_pool.py:85]: Might also wrap it in the "if name is main" clause to make sure.

**Review by cdiener — COMMENTED** [review]
Not sure what is happening. Just some suggestions that worked for me.

File: src/cobra/util/process_pool.py
Comment (cdiener) [reply src/cobra/util/process_pool.py:85]: NVM. That's the exit result. Not sure how terminate plays with a pool that has already been closed and joined though.

Comment (Midnighter) [reply src/cobra/util/process_pool.py:85]: > Not sure how terminate plays with a pool that has already been closed and joined though.

Not 100% sure either but at least it doesn't seem to fail.

Comment (Midnighter) [reply src/cobra/util/process_pool.py:85]: I'm not sure I follow you. `__name__` will always be `cobra.util.process_pool` here, won't it?

**Comment by Midnighter** [comment]
> makes me believe that either our time limits are too strict

The limits work differently now. The overall workflow has 6 hours to complete which is the maximum. The individual pytest step has 1 hour to complete. Which should be plenty considering that it finishes in less than 10 minutes normally. Maybe it's rather related to what @cdiener found in https://github.com/opencobra/cobrapy/issues/1189#issuecomment-1069498379 ?

**Comment by cdiener** [comment]
I don't think so. Those tests run fast on my machine and they did run fine on linux before the changes. The best way to validate it's indeed a deadlock is to run pytest without pytest-cov. If that is suddenly fast it's due to the deadlock.

**Comment by synchon** [comment]
So, re-triggering the 2 failed jobs fixes one case and the other one goes into the old state. I am a bit more confident about the fact that this is something related to Github's workflow queuing strategy.

**Comment by cdiener** [comment]
I did have similar behavior in micom though and I could fix it by changing the pool code.

**Comment by synchon** [comment]
Re-triggering solves the problem with the last failing job. My guess would be the runner running out of computing resources, most probably RAM.

**Comment by cdiener** [comment]
RAM could be because of the memory leak. We can try to set maxtasksperchild to something intermediate.

**Comment by synchon** [comment]
> I did have similar behavior in micom though and I could fix it by changing the pool code.

By any chance, is `micom`'s strategy matrix simpler than `cobrapy`'s?

**Comment by cdiener** [comment]
It's only 3.7-3.9 apart from that it's the same.

**Comment by synchon** [comment]
> RAM could be because of the memory leak. We can try to set maxtasksperchild to something intermediate.

After quite some community forum hops, I think it's due to the way Azure does it. Some users report that after running out of computing resources, Azure does not trigger an immediate cancel and also does not report the issue. This is a broad guess and might be something else.

What I am confused about is that if there is a memory leak, how did the previous runs on older PRs run well? One thought is that the strategy matrix previously was a bit simpler and thus the runner was able to complete it.

**Comment by cdiener** [comment]
Or they made the instances smaller. Hard to say...

**Comment by synchon** [comment]
> It's only 3.7-3.9 apart from that it's the same.

So essentially, we run: 2 versions x 3 platforms more and I think we are allotted a fixed set and amount of computing resources for each workflow.

**Comment by synchon** [comment]
> Or they made the instances smaller. Hard to say...

Highly possible. I, somehow, have had a bitter experience with Azure resources in the past. But, it might be just me.

File: src/cobra/util/process_pool.py
Comment (synchon) [inline src/cobra/util/process_pool.py:84]: Does the code in the `try` block raise exceptions? I'm thinking whether we need the `try ... finally` block at all.

File: src/cobra/util/process_pool.py
Comment (cdiener) [reply src/cobra/util/process_pool.py:85]: That's what I thought too, but no. The `__name__` variable gets filled on execution. If you have it outside of a function what you say would apply, but within a function it gets assigned when the function is called. See [here](https://github.com/micom-dev/micom/blob/d8190304695103df10c399a079f35d89d284cf6a/micom/workflows/core.py#L48). That code works fine and actually runs in jupyter etc.

File: src/cobra/util/process_pool.py
Comment (Midnighter) [reply src/cobra/util/process_pool.py:84]: I can't remember why I originally put it there (it was in the `close` method). My guess is that I really wanted to clean up the pickle file no matter what. I'm also no longer sure that these changes are needed. 🤷🏼‍♂️

Comment (synchon) [reply src/cobra/util/process_pool.py:84]: I see. 😅

File: src/cobra/util/process_pool.py
Comment (Midnighter) [reply src/cobra/util/process_pool.py:85]: So you want the `.close` and `.join` only to be called when this runs as `"__main__"` or what is supposed to be in the main guard? Currently, it always runs which doesn't seem like a bad idea?

**Comment by cdiener** [comment]
The constant need to restart the runners is becoming a bit tedious. We should try to merge this to see if it improves things or run everything with a single core in the tests to see if that avoids the issue.

**Comment by Midnighter** [comment]
Need one approving review then, please. (I prefer a real review, of course.)

File: src/cobra/util/process_pool.py
Comment (cdiener) [reply src/cobra/util/process_pool.py:85]: Probably more important for creating the pool. If that has no main guard then all subprocesses may spawn their own pool and send workloads.

**Review by cdiener — APPROVED** [review]

**Comment by Midnighter** [comment]
I didn't address your comment about the `__main__` guard yet. Will try to do that before merging.

File: src/cobra/util/process_pool.py
Comment (cdiener) [reply src/cobra/util/process_pool.py:85]: Reading up join blocks the main process until all subprocesses return. So there could be some undefined behavior if a subprocess itself calls join.

Comment (Midnighter) [reply src/cobra/util/process_pool.py:85]: Reading your linked code again. It seems to me that our functions calling the process pool should include the main guard. Do you agree?

Comment (cdiener) [reply src/cobra/util/process_pool.py:85]: Yep, agreed.

**Review by cdiener — APPROVED** [review]

**Comment by cdiener** [comment]
I did some testing and just putting the parallel tests in main guards seems to help: https://github.com/cdiener/cobrapy/tree/test/deadlock. Will restart the CI in this branch a few times to see if I can trigger a deadlock again. 

EDIT: Okay finished testing and can confirm the following.

- the tests stalling the CI are exclusively tests using the pool
- tests in pytest-cov are called outside of the main process (so main guards in the code base itself will lead to tests failing)
- having the pool join/close *and* having the tests in main guards does fix the issue

**Comment by cdiener** [comment]
Should I try to open a new PR with the mentioned fixes or merge them in here @Midnighter?

**Comment by akaviaLab** [comment]
I think this can be closed now.

**Pull request closed without merging**
